### ISSUE TITLE
helm strips trailing slash with `read-directory-name`

### ISSUE THREAD

**Issue by marsam** [emacs-helm/helm#1083]
Hi:
In first place, thanks for creating helm! I can't believe how I lived without it.

Now, I have a simple function to set `default-directory` before calling the `compile` command.

```elisp
(defun marsam/compile-in-directory (directory)
  "Execute `compile' and set DIRECTORY as `default-directory'."
  (interactive (list (if current-prefix-arg (read-directory-name "Directory: ") default-directory)))
  (let ((default-directory directory))
    (call-interactively #'compile)))
```
But helm removes the trailing slash when selecting a directory, so the `compile` command uses `/path/to/some/dir` (withtout the trailing slash) and internally sets `default-directory` to `/path/to/some/`

A way to reproduce could be calling `check-directory-trailing-slash` within a directory:

```elisp
(require 'cl-lib)

;; Stolen from `s.el'
(defun string-ends-with-p (suffix s &optional ignore-case)
  (let ((start-pos (- (length s) (length suffix))))
    (and (>= start-pos 0)
         (eq t (compare-strings suffix nil nil
                                s start-pos nil ignore-case)))))

(defun check-directory-trailing-slash (directory)
  "Read DIRECTORY and check if ends with slash character."
  (interactive (list (read-directory-name "Directory: ")))
  (message "directory -> %s" directory)
  (cl-assert (string-ends-with-p "/" directory)))
```

![](https://i.imgur.com/0Y5MSFO.gif)

My helm configuration (using [`use-package`](https://github.com/jwiegley/use-package))
```elisp
(use-package helm
  :ensure t
  :init (helm-mode 1)
  :diminish (helm-mode))
```

I recently moved from ido to helm, in `ido-mode` after calling `ido` you can press `C-f` to switch to a plain mode (without `ido-mode`) is there a similar command in helm? I can't find anything in the documentation.

Thanks in advance

**Comment by thierryvolpiatto** [comment]
Mario Rodas <notifications@github.com> writes:

> Hi:
> In first place, thanks for creating helm! I can't believe how I lived without it.

Thanks.

> Now, I have a simple function to set default-directory before calling the compile command.
>
> (defun marsam/compile-in-directory (directory)
>   "Execute `compile' and set DIRECTORY as `default-directory'."
>   (interactive (list (if current-prefix-arg (read-directory-name "Directory: ") default-directory)))
>   (let ((default-directory directory))
>     (call-interactively #'compile)))
>
> But helm removes the trailing slash when selecting a directory,

Should be fixed now by 6a7f8c9.
BTW to ensure you always use a trailing slash on a directory (specially
when setting default-directory) it is safer to use file-name-as-directory.

> I recently moved from ido to helm, in ido-mode after calling ido you
> can press C-f to switch to a plain mode (without ido-mode) is there a
> similar command in helm? I can't find anything in the documentation.

No.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by marsam** [comment]
> Should be fixed now by 6a7f8c9.

Thanks!

>> is there a similar command in helm? I can't find anything in the documentation.

> No.

ok, no problem.
